### PR TITLE
added startsWith and endsWith

### PR DESCRIPTION
**Pull request by martinflory**
Ok...
- rolled back to .js, and used the underscore-string functions for `startsWith` and `endsWith`.
- made the linter work with .js (for this project for now).
- added some tests for this functionality.

Our linter module is private, we should probably remove it from here, or make it public? If we think using it for Baseline (either .js or .es) is a good idea, then maybe make public. What do you think @pepoviola @dgbeck ?

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -2,9 +2,14 @@
   "name": "matches-where-query",
   "version": "1.1.5",
   "main": "matchesWhereQuery.js",
+  "scripts": {
+    "test": "mocha tests/test.js --reporter spec"
```
Comment (pepoviola) [inline package.json:6]: We should setup the CI to run this test, and also we can add more test to cover all cases.
Thx!

File: matchesWhereQuery.js [hunk 2]
```diff
@@ -3,45 +3,73 @@ var _ = require( 'underscore' );
 module.exports = function( object, whereQuery ) {
 	var whereQueryKeys = _.keys( whereQuery );
 	for( var i = 0; i < whereQueryKeys.length; i++ ) {
-		var thisKey = whereQueryKeys[ i ];
-		var thisKeyQuery = whereQuery[ thisKey ];
-		var thisObjectValue = object[ thisKey ];
+		var attribute = whereQueryKeys[ i ];
```
Comment (pepoviola) [inline matchesWhereQuery.js]: I think we can use another name for this `var`. I personally prefer `key` over `attribute`. What did you think @dgbeck .
Thx!

**Review by pepoviola — COMMENTED** [review]
Looks good @martinflory, small comments inline.
Thx!

**Comment by martinflory** [comment]
Oh, I just found a case in which it tries to apply the rules for es6 in files to be parsed as es5. :( So eslint might not be as easy as I thought.

File: matchesWhereQuery.js [hunk 3]
```diff
@@ -3,45 +3,73 @@ var _ = require( 'underscore' );
...
-			if( thisKeyQuery !== object[ thisKey ] ) return false;
+			if( queryAttribute !== objectAttribute ) return false;
 		}
 	}
-
 	return true;
 };
+
+function _startsWith( str, starts ) {
+	str = _makeString( str );
```
Comment (dgbeck) [inline matchesWhereQuery.js]: we already know this is a string, because we check it on li nne 38

File: matchesWhereQuery.js [hunk 4]
```diff
@@ -3,45 +3,73 @@ var _ = require( 'underscore' );
...
+function _endsWith( str, ends ) {
+	str = _makeString( str );
+	ends = String( ends );
+	var position = str.length - ends.length;
+
+	return position >= 0 && str.indexOf( ends, position ) === position;
+}
+
+function _makeString( object ) {
+	if( ! object ) return '';
```
Comment (dgbeck) [inline matchesWhereQuery.js]: same as above, we know this is a string

File: matchesWhereQuery.js [hunk 5]
```diff
@@ -3,45 +3,73 @@ var _ = require( 'underscore' );
...
+				if( ! _.isArray( queryAttribute.value ) || queryAttribute.value.length !== 2 ) {
 					throw new Error( 'Value supplied for isBetween comparator must be an array [ min, max ]' );
 				}
 
-				if( thisObjectValue < thisKeyQuery.value[ 0 ] || thisObjectValue > thisKeyQuery.value[ 1 ] ) return false;
+				if( objectAttribute < queryAttribute.value[ 0 ] || objectAttribute > queryAttribute.value[ 1 ] ) return false;
+				break;
+			case 'startsWith':
+				// eslint-disable-next-line max-len
+				if( ! _.isString( objectAttribute ) || ! _.isString( queryAttribute.value ) ) throw new Error( 'For startsWith comparator both the object attribute and the query value must be strings.' );
```
Comment (dgbeck) [inline matchesWhereQuery.js:39]: i think we might want to consider the `null` case here (as well as with `endsWith`). if the `objectAttribute` is `null` that should be a non-match

File: matchesWhereQuery.js (hunk 2 above)
Comment (dgbeck) [reply matchesWhereQuery.js]: seems pretty similar to me, no preference, my head spins for all of these names!

**Comment by dgbeck** [comment]
hi @martinflory comments inline, thx!

File: matchesWhereQuery.js (hunk 2 above)
Comment (pepoviola) [reply matchesWhereQuery.js]: Oks, sounds good. 👍

Comment (martinflory) [reply matchesWhereQuery.js]: Talked to Javi and changed it to attributeName. :)

File: matchesWhereQuery.js (hunk 3 above)
Comment (martinflory) [reply matchesWhereQuery.js]: So true, thanks! Fixed!

File: matchesWhereQuery.js (hunk 4 above)
Comment (martinflory) [reply matchesWhereQuery.js]: same! fixed, thx!

File: matchesWhereQuery.js (hunk 5 above)
Comment (martinflory) [reply matchesWhereQuery.js:39]: Good one, its done, and tested! thanks

File: package.json (hunk 1 above)
Comment (martinflory) [reply package.json:6]: Yes, thanks! Its working fine now and with the tests. :)

**Comment by pepoviola** [comment]
Awesome, I like the test!! ping @dgbeck  to merge and publish.
Thx!

**Comment by dgbeck** [comment]
awesome, thx!

File: matchesWhereQuery.js [hunk 6]
```diff
@@ -3,45 +3,69 @@ var _ = require( 'underscore' );
...
-			if( thisKeyQuery !== object[ thisKey ] ) return false;
+			if( queryAttribute !== objectAttribute ) return false;
 		}
 	}
-
 	return true;
 };
+
+function _startsWith( str, starts ) {
+	starts = String( starts );
```
Comment (dgbeck) [inline matchesWhereQuery.js]: @martinflory we are already checking that this is a string... why cast it here?

File: matchesWhereQuery.js [hunk 7]
```diff
@@ -3,45 +3,69 @@ var _ = require( 'underscore' );
...
+	return str.lastIndexOf( starts, 0 ) === 0;
+}
+
+function _endsWith( str, ends ) {
+	ends = String( ends );
```
Comment (dgbeck) [inline matchesWhereQuery.js]: same here

**Comment by dgbeck** [comment]
@martinflory one comment then good to merge, thx!

File: matchesWhereQuery.js (hunk 6 above)
Comment (martinflory) [reply matchesWhereQuery.js]: Yes! You're right. thanks

File: matchesWhereQuery.js (hunk 7 above)
Comment (martinflory) [reply matchesWhereQuery.js]: Yes! You're right. thanks

**Comment by martinflory** [comment]
All sorted! thanks for you corrections and suggestions @dgbeck @pepoviola! Merging now.

**Pull request merged**